visions/abi-stability.md

```diff
@@ -0,0 +1,248 @@
+18.04 will likely run on Ubuntu 18.10, but might not work on Ubuntu
+22.04).  Even if all the libraries you are using are themselves ABI
+stable, the versions of libraries installed by different distributions
+may differ---distribution X may have a newer version of Glibc, but an
+older version of (say) zlib, while distribution Y has an older Glibc
+but a newer zlib.  And some distributions make different choices about
+what "the system C library" or "the system C++ library" might be (this
+isn't limited to low-level libraries either---similar choices might
+happen for much higher layers of the system, such as which window
+system to use, whether to use MIT Kerberos or Heimdall, and so on).
+
+The upshot is that it is difficult to distribute binary programs for
+Linux, without doing what the Swift Static SDK for Linux does and
+simply not relying on external dynamic libraries.  This difficulty has
+also led to the creation of systems like Snaps and Flatpak, which let
+an application bundle all its dependencies in such a way that they
+won't interfere with other installed applications.  It is also a
+factor in the rise in popularity of OCI images, particularly for
+server applications.
+
+There are some system features intended to support ABI stability at
+the library level, including versioned shared object names and symbol
+versioning.  These get used by ABI stable libraries like Glibc such
+that it's generally possible to run a program that depends on such a
+library on the version it was built for _or any newer version_.  This
+does make it possible to ship binaries that will run on a wide variety
+of Linux systems by ensuring that your program is linked against the
+oldest version(s) of dynamic libraries that you care about, _provided
+those libraries are themselves ABI stable_.  Python's "manylinux"
+package builds take this approach, but have to restrict themselves as
+a result to a small subset of libraries that are known to be ABI
+stable.
+
+## Goals
+
+* Swift should support existing platform ABI stability mechanisms,
+  so that the language is a good citizen when writing native software.
+
+  - `@availability` should be well-defined for Windows; it should be
+    possible to annotate Windows APIs to say e.g. that they were
+    introduced in Windows 10 version 1709.  Swift should automatically
+    generate the code to weakly reference functions it imports where
+    necessary.
+
+  - `@availability` as a system-wide concept doesn't make as much
+    sense on non-Apple UNIX/UNIX-like platforms where it's possible to
+    install different versions of many standard packages.  This is
+    especially true on Linux where there are multiple "distributions",
+    so the major versions of packages in major versions of those
+    distributions are fixed, but one distribution's set of packages
+    may bear little relation to another's.  In this case,
+    `@availability` annotations could perhaps be used for the system C
```

Comment:
It would be great to have an actual Swift-version availability annotation that doesn't map back to an OS version.

Reply:
Agreed.